sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+      builder.addIfNotNull(DisplayData.item("filter", filter));
+      builder.addIfNotNull(DisplayData.item("numSplits", numSplits));
+    }
+
+  }
+
+  private static class BoundedMongoDbSource extends BoundedSource {
+
+    private final String uri;
+    private final String database;
+    private final String collection;
+    @Nullable
+    private final String filter;
+    private final int numSplits;
+
+    public BoundedMongoDbSource(String uri, String database, String collection, String filter,
+                                int numSplits) {
+      this.uri = uri;
+      this.database = database;
+      this.collection = collection;
+      this.filter = filter;
+      this.numSplits = numSplits;
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(String.class);
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(uri, "uri");
+      Preconditions.checkNotNull(database, "database");
+      Preconditions.checkNotNull(collection, "collection");
+    }
+
+    @Override
+    public boolean producesSortedKeys(PipelineOptions options) {
+      return false;
+    }
+
+    @Override
+    public BoundedReader createReader(PipelineOptions options) {
+      return new BoundedMongoDbReader(this);
+    }
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+      long estimatedSizeBytes = 0L;
+
+      MongoClient mongoClient = new MongoClient();
+      MongoDatabase mongoDatabase = mongoClient.getDatabase(database);
+      MongoCollection mongoCollection = mongoDatabase.getCollection(collection);
+
+      // get the Mongo collStats object
+      // it gives the size for the entire collection
+      BasicDBObject stat = new BasicDBObject();
+      stat.append("collStats", collection);
+      Document stats = mongoDatabase.runCommand(stat);
+      estimatedSizeBytes = Long.valueOf(stats.get("size").toString());
+      long avgSize = Long.valueOf(stats.get("avgObjSize").toString());
+      if (filter != null && !filter.isEmpty()) {
+        // as the user is using a filter, we improve the estimated size
+        // by counting number of document returned by the filter
+        // and using the average object size from collStats
+        try {
+          estimatedSizeBytes = getFilterResultByteSize(mongoCollection, avgSize);
+        } catch (Exception e) {
+          LOGGER.warn("Can't estimate size", e);
+        }
+      }
+      return estimatedSizeBytes;
+    }
+
+    private long getFilterResultByteSize(MongoCollection mongoCollection, long avgSize)
+        throws Exception {
+      // count the number of documents selected by the filter and estimate the size of the
+      // filter using the average object size * the number of documents
+      long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
```

Comment:
Same question as last time: does this actually execute the query against the whole database? It sounds like a sequential bottleneck that could defeat the parallelism you're achieving in splitIntoBundles.